### ISSUE TITLE
deepImmutablesDetection: 1. broken with optional (bug), 2. no interop with non @Immutable immutables (improvement), 3. documentation missing

### ISSUE THREAD

**Issue by vorburger** [immutables/immutables#471]
Based on a more general version of what I was going on about in https://groups.google.com/forum/#!topic/immutables/TXkXlXEcn4g, how about generating _alternative setters with automated inlined versions for any types having constructor/s?_  So given:

```
@Immutable
interface class Something {
    Thing thing();
}
```

IFF the `Thing` class (or if its an interface or abstract class then its its Immutable\* implementation gen. by Immutables) has a constructor _(or an Immutables `of()` like static factory method; hell even if it had several of either of 'em...)_ then `ImmutableSomething`, in addition to the existing standard `Builder thing(Thing thing)` setter would have a _(several)_ `Builder thing(<Thing cstr args>)` with an implementation body which does `new Thing(<Thing cstr args from setter args>)` (or `Thing.of(..)` - you know what I mean).

See https://groups.google.com/forum/#!topic/immutables/TXkXlXEcn4g for more background re. the idea.

That would be so cool. I think.

**Comment by elucash** [comment]
Hi @vorburger Sorry I didn't have time to make a proper answer on a mailing list. But I just want to let you know that this/similar feature is already implemented: use `deepImmutablesDetection=true` style attribute and your other objects has to have constructors generated, so the builder will have shortcut init methods.

https://github.com/immutables/immutables/blob/master/value-fixture/src/org/immutables/fixture/deep/Canvas.java

```java
@Value.Style(deepImmutablesDetection = true, depluralize = true)
public interface Canvas {

  @Value.Modifiable
  @Value.Immutable
  public interface Color {
    @Value.Parameter
    double red();

    @Value.Parameter
    double green();

    @Value.Parameter
    double blue();
  }

  @Value.Modifiable
  @Value.Immutable
  public interface Line {
    List<Point> points();

    Color color();
  }

  @Value.Modifiable
  @Value.Immutable
  public interface Point {
    @Value.Parameter
    int x();

    @Value.Parameter
    int y();
  }

  default void use() {
    ImmutableLine line = ImmutableLine.builder()
        .color(0.9, 0.7, 0.4)
        .addPoint(1, 2)
        .addPoint(2, 3)
        .addPoint(4, 5)
        .build();

    ImmutableColor color = line.color();

    ImmutableLine.builder()
        .addAllPoints(line.points())
        .color(color)
        .build();
  }
}
```

**Comment by vorburger** [comment]
@elucash OMG - wow. Yes, deepImmutablesDetection is very interesting. Thank you! I've immediately tried using it in a project; and have some feedback, please take this as constructively as I mean it (immutables.org is amazing!) :

1. Sorry I didn't RTFM .. oh wait, this [isn't mentioned anywhere in the fM](https://github.com/immutables/immutables.github.io/search?utf8=%E2%9C%93&q=deepImmutablesDetection), yet... ;-) Perhaps rename this issue into "deepImmutablesDetection documentation missing" ? Contribution welcome? How about somewhere in that "Patterns" chapter, perhaps as 2nd point, right after you introduce "Wrapper types" there?

2. It doesn't seem to work for optional attributes yet, could that be a bug, or am I just really dumb? Take for example your `Canvas` above - works great, thank you very much. Now make that `color()` of your Line an `Optional<Color>` ... and now `ImmutableLine.builder().color(0.9, 0.7, 0.4)` doesn't work anymore, even though there still is a `Builder color(Canvas.Color color)` initializer - that doesn't seem right, like a bug, to me, do you agree? Do you want a separate bug about this? This point is the reason why I can't actually switch yet to use deepImmutablesDetection as is.

3. In deepImmutablesDetection, as its documentation actually clearly does say, _"each attribute is being analized [analyzed, sic!] and if it [is] discovered to be an `Value.Immutable` object"_, only works for [Immutables.org](url) immutable types. Would you consider expanding this to be a little more open and generic, similar to what I had initially outlined above? For interop with value classes from other libraries, this would be great. Not just making this up, but I actually have this need in a project where this would be nice.. we have some class Uuid, not gen. by [Immutables.org](url) (not hand written either, but by... some other kinda similar tool, YANG code gen. - whatever) - which is an immutable value object, and it has a public somelib.Uuid(java.lang.String _value) constructor - and that's even annotated with @[java.beans.ConstructorProperties](https://docs.oracle.com/javase/7/docs/api/java/beans/ConstructorProperties.html)("value"), if that's of any use. Now it would be... "nice" (great!) if an @Immutable interface Thing { somelib.Uuid uuid() } could have a uuid(String value) as well (currently it does not, as the doc says; I've actually tried it out). Do you want a separate enhancement issue for this point?

**Comment by elucash** [comment]
Thank you for the kind words!
1. Any contributions very are welcome, especially for the documentation. I'm currently busy writing documentation/tutorial for the encodings (#363), so can get to deep-detection only after.
2. We can make it more consistent. I'm not sure if there are any hard reasons to not support it
3. Maybe we can read free-form classes, but it would require some hints like `ConstructoProperties` or some form of inference. While not too hard by itself, but it may be complicated by a lot of issues along the way. Probably, some help/contributions will be needed for this. It's better, probably, to not create additional issues, just rename the title

**Comment by elucash** [comment]
added Optional support in version 2.3.8 (currently propagating to Central)

**Comment by mghildiy** [comment]
is this issue is still open, I would like to work on it.

**Comment by elucash** [comment]
@mghildiy Please, let us know what is specific issue is, here or as another ticket. This ticket raised a number of concerns as sparked some discussion, and after that, it might have lost some relevance, so it would be useful to refresh and describe again any remaining issues. Thank you!

**Comment by mghildiy** [comment]
@elucash , just wanted to contribute to this project. So if there are any other issues, I can look into them.
